inc/Engine/CriticalPath/RESTGenerate.php

```diff
@@ -0,0 +1,392 @@
+	 * @return array
+	 */
+	protected function send_generation_request( $post_url, $post_type ) {
+		$response = wp_remote_post(
+			self::API_URL,
+			[
+				// This filter is documented in inc/Engine/CriticalPath/CriticalCSSGeneration.php.
+				'body' => apply_filters(
+					'rocket_cpcss_job_request',
+					[
+						'url' => $post_url,
+					],
+					[
+						'url'  => $post_url,
+						'type' => $post_type,
+					]
+				),
+			]
+		);
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+
+		if ( in_array( $response_code, [ 400, 404 ], true ) ) {
+			$data = json_decode( wp_remote_retrieve_body( $response ) );
+
+			// translators: %1$s = post URL.
+			$error = sprintf( __( 'Critical CSS for %1$s not generated.', 'rocket' ), $post_url );
+
+			if ( isset( $data->message ) ) {
+				// translators: %1$s = error message.
+				$error .= ' ' . sprintf( __( 'Error: %1$s', 'rocket' ), $data->message );
+			}
+
+			return $this->return_array_response(
+				false,
+				'cpcss_generation_failed',
+				// translators: %s = post URL.
+				$error,
+				400
+			);
+		}
+
+		if ( 200 !== $response_code ) {
+			return $this->return_array_response(
+				false,
+				'cpcss_generation_failed',
+				// translators: %s = post URL.
+				sprintf( __( 'Critical CSS for %1$s not generated. Error: The API returned an invalid response code.', 'rocket' ), $post_url ),
+				$response_code
+			);
+		}
+
+		$data = json_decode( wp_remote_retrieve_body( $response ), true );
+
+		if ( ! isset( $data['data']['id'] ) ) {
+			return $this->return_array_response(
+				false,
+				'cpcss_generation_failed',
+				// translators: %s = post URL.
+				sprintf( __( 'Critical CSS for %1$s not generated. Error: The API returned an empty response.', 'rocket' ), $post_url ),
+				400
+			);
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Checks CPCSS job status for the given ID.
+	 *
+	 * @since 3.6
+	 *
+	 * @param  int    $cpcss_job_id The CPCSS Job ID.
+	 * @param  int    $post_id      The post ID.
+	 * @param  string $post_url     The post URL.
+	 * @param  string $post_type    The post type.
+	 * @return array
+	 */
+	protected function check_cpcss_job_status( $cpcss_job_id, $post_id, $post_url, $post_type ) {
+		$job_data = $this->get_critical_path( $cpcss_job_id );
+
+		if ( in_array( (int) $job_data->status, [ 400, 404 ], true ) ) {
+			return $this->on_job_error( $post_id, $post_url, $job_data );
+		}
+
+		if ( isset( $job_data->data->state ) && 'complete' !== $job_data->data->state ) {
+			return $this->return_array_response(
+				true,
+				'cpcss_generation_pending',
+				// translators: %s = post URL.
+				sprintf( __( 'Critical CSS for %s in progress.', 'rocket' ), $post_url ),
+				200
+			);
+		}
+
+		if ( isset( $job_data->data->state, $job_data->data->critical_path ) && 'complete' === $job_data->data->state ) {
+			return $this->on_job_success( $post_id, $post_url, $post_type, $job_data->data->critical_path );
+		}
+	}
```

Comment:
I assume that CPCSS API always have a `data->state` returned. Otherwise will return a `400 or 404` error. But we can return a generic error as a fallback.

Reply:
Let's validate that assumption @crystinutzaa.

Reply:
@hellofromtonya I think this Wiki will be enough:
https://github.com/wp-media/criticalpath-css-generator/wiki/API

Reply:
Agreed.

Reply:
In that case, it can be simplified:
```php
	protected function check_cpcss_job_status( $cpcss_job_id, $post_id, $post_url, $post_type ) {
		$job_data = $this->get_critical_path( $cpcss_job_id );

		if ( in_array( (int) $job_data->status, [ 400, 404 ], true ) ) {
			return $this->on_job_error( $post_id, $post_url, $job_data );
		}

		if ( 'complete' !== $job_data->data->state ) {
			return $this->return_array_response(
				true,
				'cpcss_generation_pending',
				// translators: %s = post URL.
				sprintf( __( 'Critical CSS for %s in progress.', 'rocket' ), $post_url ),
				200
			);
		}

		return $this->on_job_success( $post_id, $post_url, $post_type, $job_data->data->critical_path );
	}
```

Reply:
Thank you 🙏